pgxn/neon/pagestore_smgr.c

```diff
@@ -126,35 +127,335 @@ static UnloggedBuildPhase unlogged_build_phase = UNLOGGED_BUILD_NOT_IN_PROGRESS;
+prefetch_set_unused(uint64 ring_index)
+{
+	PrefetchRequest *slot = &MyPState->prf_buffer[ring_index % READ_BUFFER_SIZE];
+
+	if (slot->status == PRFS_UNUSED)
+		return;
+
+	Assert(slot->status == PRFS_RECEIVED || slot->status == PRFS_TAG_REMAINS);
+	Assert(ring_index >= MyPState->ring_last &&
+		   ring_index < MyPState->ring_unused);
+
+	if (slot->status == PRFS_RECEIVED)
+	{
+		pfree(slot->response);
+		slot->response = NULL;
+
+		MyPState->n_responses_buffered -= 1;
+		MyPState->n_unused += 1;
+	}
+	else
+	{
+		Assert(slot->response == NULL);
+	}
+
+	prfh_delete(MyPState->prf_hash, slot);
+
+	/* clear all fields */
+	MemSet(slot, 0, sizeof(PrefetchRequest));
+	slot->status = PRFS_UNUSED;
+
+	/* run cleanup if we're holding back ring_last */
+	if (MyPState->ring_last == ring_index)
+		prefetch_cleanup();
+}
+
+static void
+prefetch_do_request_ext(PrefetchRequest *slot, bool *force_latest, XLogRecPtr *force_lsn)
+{
+	NeonGetPageRequest request = {
+		.req.tag = T_NeonGetPageRequest,
+		.req.latest = false,
+		.req.lsn = 0,
+		.rnode = slot->buftag.rnode,
+		.forknum = slot->buftag.forkNum,
+		.blkno = slot->buftag.blockNum,
+	};
+	slot->effective_request_lsn = prefetch_lsn;
+	
+	if (force_lsn && force_latest)
+	{
+		request.req.lsn = *force_lsn;
+		request.req.latest = *force_latest;
+	}
+	else
+	{
+		request.req.lsn = neon_get_request_lsn(
+			&request.req.latest,
+			slot->buftag.rnode,
+			slot->buftag.forkNum,
+			slot->buftag.blockNum
+		);
+	}
+
+	MyPState->n_requestes_inflight++;
+
+	page_server->send((NeonRequest *) &request);
+	Assert(slot->response == NULL);
+	slot->status = PRFS_REQUESTED;
+}
+
+/*
+ * prefetch_register_buffer() - register and prefetch buffer
+ *
+ * Register that we may want the contents of BufferTag in the near future.
+ */
+static uint64
+prefetch_register_buffer(BufferTag tag)
+{
+	return prefetch_register_buffer_ext(tag, NULL, NULL);
+}
+
+static uint64
+prefetch_register_buffer_ext(BufferTag tag, bool *force_latest, XLogRecPtr *force_lsn)
+{
+	int		index;
+	bool	found;
+	uint64	ring_index;
+	PrefetchRequest *slot;
+	PrfHashEntry *entry;
+
+	entry = prfh_lookup(MyPState->prf_hash, (PrefetchRequest *) &tag);
+
+	if (entry != NULL)
 	{
-		NeonResponse *resp = page_server->receive();
+		slot = entry->slot;
+		ring_index = slot->my_ring_index;
+		index = (ring_index % READ_BUFFER_SIZE);
+		Assert(slot == &MyPState->prf_buffer[index]);
 
-		pfree(resp);
+		Assert(slot->status != PRFS_UNUSED);
+		Assert(BUFFERTAGS_EQUAL(slot->buftag, tag));
+
+		/* we received a prefetch for a page that was recently read */
+		if (slot->status == PRFS_TAG_REMAINS)
+		{
+			prefetch_set_unused(ring_index);
+		}
+		else
+		{
+			n_prefetch_dupes++;
+			return ring_index;
+		}
 	}
-	n_prefetched_buffers = 0;
-	n_prefetch_responses = 0;
+
+	ring_index = MyPState->ring_unused;
+	index = (ring_index % READ_BUFFER_SIZE);
+	slot = &MyPState->prf_buffer[index];
+
+	if (MyPState->ring_last == (MyPState->ring_unused - READ_BUFFER_SIZE))
```

Comment:
```suggestion
	/*
	 * If we prefetch queue is full, we need to make room by clearing the oldest slot. If the oldest slot
	  * holds a buffer that was already received, we have to throw it away; we fetched the page
	  * unnecessarily in that case. If the oldest slot holds a request that we haven't received a response for
	  * yet, we have to wait for the response to that before we can continue. We might not have even flushed
	  * the request to the pageserver yet, it might be just sitting in the output buffer. In that case, we flush it
	  * and wait for the response. (We could decide not to send it, but it's hard to abort when the request is
	  * already in the output buffer)
	  */
	if (MyPState->ring_last == (MyPState->ring_unused - READ_BUFFER_SIZE))
```
Did I understand this correctly? Maybe we should not call `page_server->send()` on the requests until just before we flush them? That way, we could decide to just drop the unsent request here. It probably doesn't happen very often to matter much, though.